tempesta_fw/http2/huffman.c

```diff
@@ -0,0 +1,1385 @@
+
+#define GET_OCTETS(tail)			    \
+do {						    \
+	unsigned int space = 0; 		    \
+	if (n < Word_Size - 1) {		    \
+		space = (Bit_Capacity - 8) - n * 8; \
+		n = Word_Size - 1;		    \
+		current -= space;		    \
+	}					    \
+	if ((uintptr_t) src & 1) {		    \
+		c = Bit_Join8(c, * src++);	    \
+		space += 8;			    \
+	}					    \
+	current += (Word_Size - 1) * 8; 	    \
+	n -= Word_Size - 1;			    \
+	GET_CASCADE();				    \
+} while (0)
+
+#define GET_FIRST_CASCADE()						   \
+do {									   \
+	c = 0;								   \
+	if ((uintptr_t) src & 1) {					   \
+		c = * src++;						   \
+		space += 8;						   \
+	}								   \
+	if ((uintptr_t) src & 2) {					   \
+		if (space <= Bit_Capacity - 16) {			   \
+			space += 16;					   \
+			c = Bit_Join(c, 16,				   \
+				     Big16(* (uint16_t *) src));	   \
+			src += 2;					   \
+L0:									   \
+			if (space <= Bit_Capacity - 32) {		   \
+				space += 32;				   \
+				c = Bit_Join(c, 32,			   \
+					     Big32(* (uint32_t *) src));   \
+				src += 4;				   \
+			}						   \
+			if (space <= Bit_Capacity - 16) {		   \
+				space += 16;				   \
+				c = Bit_Join(c, 16,			   \
+					     Big16(* (uint16_t *) src));   \
+				src += 2;				   \
+			}						   \
+		}							   \
+		if (space <= Bit_Capacity - 8) {			   \
+			c = Bit_Join8(c, * src++);			   \
+		}							   \
+	}								   \
+	else if (space) {						   \
+		goto L0;						   \
+	}								   \
+	else {								   \
+		c = BigWide(						   \
+			Bit_Join((uintptr_t) * (uint32_t *) (src + 4), 32, \
+					     * (uint32_t *) src)	   \
+		);							   \
+		src += Word_Size;					   \
+	}								   \
+} while (0)
+
+#define GET_FIRST(tail) 				\
+do {							\
+	current = Bit_Capacity - HT_NBITS;		\
+	if (((uintptr_t) src & (Word_Size - 1)) == 0) { \
+		if (n < Word_Size) {			\
+			goto Z0;			\
+		}					\
+		c = BigWide(* (uintptr_t *) src);	\
+		src += Word_Size;			\
+	}						\
+	else {						\
+		unsigned int space = 0; 		\
+		if (n < Word_Size) {			\
+Z0:							\
+			space = Bit_Capacity - n * 8;	\
+			n = Word_Size;			\
+			current -= space;		\
+		}					\
+	}						\
+	GET_FIRST_CASCADE();				\
+	n -= Word_Size; 				\
+} while (0)
+
+#else
+
+#ifdef Platform_Alignment
+
+#define GET_OCTETS(tail)			 \
+	if (current != -tail || 		 \
+	    (uintptr_t) src & (Word_Size - 1) || \
+	    n < Word_Size)			 \
+		GET_OCTETS_BY_8(tail);		 \
+	else					 \
+		GET_UWIDE(tail)
+
+#define GET_FIRST(tail) 				\
+	if (((uintptr_t) src & (Word_Size - 1)) == 0 && \
+	    n >= Word_Size)				\
+		GET_UWIDE(tail);			\
+	else						\
+		GET_FIRST_BY_8(tail)
+
+#else
+
+#define GET_OCTETS(tail)		       \
+	if (current != -tail || n < Word_Size) \
+		GET_OCTETS_BY_8(tail);	       \
+	else				       \
+		GET_UWIDE(tail)
+
+#define GET_FIRST(tail) 	     \
+	if (n >= Word_Size)	     \
+		GET_UWIDE(tail);     \
+	else			     \
+		GET_FIRST_BY_8(tail)
+
+#endif
+
+#endif
+
+unsigned int
+huffman_decode(const char *__restrict source, char *__restrict dst, uintptr_t n)
```

Comment:
The algorithm looks good: `ht_decode` is relatively small (about 4*848 bytes, which is smaller that 4 pages of [huff_decode_table](https://github.com/nghttp2/nghttp2/blob/master/lib/nghttp2_hd_huffman_data.c#L96) and handles the most probable 1/2 of characters in single step. The algorithms is somewhat similar to http://fastcompression.blogspot.com/2015/07/huffman-revisited-part-3-depth-limited.html

However, I'd love to see the proof that the algorithm is really fast, so please write a benchmark test comparing it with [nghttp2_hd_huff_decode()](https://github.com/nghttp2/nghttp2/blob/master/lib/nghttp2_hd_huffman.c#L198).

A good readings about fast Huffman decoders:
https://www.ifi.uzh.ch/dam/jcr:ffffffff-82b7-d340-0000-000026dd8fb6/Prefix.pdf
http://citeseerx.ist.psu.edu/viewdoc/download?doi=10.1.1.93.9447&rep=rep1&type=pdf
http://fastcompression.blogspot.com/2015/07/huffman-revisited-part-1.html
http://fastcompression.blogspot.com/2015/07/huffman-revisited-part-2-decoder.html
http://fastcompression.blogspot.com/2015/07/huffman-revisited-part-3-depth-limited.html
http://fastcompression.blogspot.com/2015/10/huffman-revisited-part-4-multi-bytes.html
http://fastcompression.blogspot.com/2015/10/huffman-revisited-part-5-combining.html

However I don't see much sense in implementation of the multi-stream decoder since we need to feed the data to our parser and the data chunks aren't so big.